src/Map.jl

```diff
@@ -0,0 +1,476 @@
+"""
+    AbstractMap{D,CoD}
+
+Abstract type for maps between elements of sets.
+Every map has a domain of type `D`, which is the set to which inputs
+belong and a codomain `CoD`, which is the set to which outputs belong. Note
+that maps are not required to be total. That is, the true (co)domain may be
+a subset of the provided (co)domain.
+
+Every new map type must implement [`domain`](@ref) and [`codomain`](@ref) and be
+callable. Optionally, [`inv`](@ref), [`linv`](@ref), and [`rinv`](@ref) may be
+used to specify a map that is a true, left-, or right-inverse, respectively.
+
+Maps can be composed in the same way as functions using the `∘` command. See
+[`CompositeMap`](@ref).
+"""
+abstract type AbstractMap{D,CoD} end
+
+function show(io::IO, mime::MIME"text/plain", m::AbstractMap)
+    print(io, "$(typeof(m).name): 𝑀 ⟶ 𝑁\n",
+              "𝑀 = $(repr(mime, domain(m)))\n",
+              "𝑁 = $(repr(mime, codomain(m)))")
+end
+
+"""
+    domain(m::AbstractMap)
+
+Get the manifold to which inputs to the map `m` belong. By default, this is
+assumed to be stored in the field `m.domain`.
+"""
+domain(m::AbstractMap) = m.domain
+domain(::AbstractMap{M}) where {M<:Euclidean} = M()
+
+"""
+    codomain(m::AbstractMap)
+
+Get the manifold to which outputs to the map `m` belong. By default, this is
+assumed to be stored in the field `m.codomain`.
+"""
+codomain(m::AbstractMap) = m.codomain
+codomain(::AbstractMap{M,N}) where {M,N<:Euclidean} = N()
+
+@doc doc"""
+    inv(f::AbstractMap)
+
+For a _total_ map $f \colon M \to M$, return its true inverse $g$, such that
```

Comment:
Shouldn't `inv` work for all invertible maps (not only those for which domain and codomain are equal?)